room/compiler-processing/src/main/java/androidx/room/compiler/processing/ksp/KSTypeExt.kt

```diff
@@ -0,0 +1,78 @@
+/*
+ * Copyright 2020 The Android Open Source Project
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package androidx.room.compiler.processing.ksp
+
+import com.squareup.javapoet.ClassName
+import com.squareup.javapoet.ParameterizedTypeName
+import com.squareup.javapoet.TypeName
+import org.jetbrains.kotlin.ksp.symbol.KSClassifierReference
+import org.jetbrains.kotlin.ksp.symbol.KSDeclaration
+import org.jetbrains.kotlin.ksp.symbol.KSName
+import org.jetbrains.kotlin.ksp.symbol.KSType
+import org.jetbrains.kotlin.ksp.symbol.KSTypeReference
+
+// catch-all type name when we cannot resolve to anything.
+private val UNDEFINED = ClassName.get("androidx.room.compiler.processing.kotlin.error", "Undefined")
+
+/**
+ * Turns a KSTypeReference into a TypeName
+ *
+ * We try to achieve this by first resolving it and iterating.
+ * If some types cannot be resolved, we do a best effort name guess from the KSTypeReference's
+ * element.
+ */
+internal fun KSTypeReference?.typeName(): TypeName {
+    if (this == null) {
+        return UNDEFINED
+    }
+    val resolved = resolve()
+    return resolved?.typeName() ?: fallbackClassName()
```

Comment:
Can be written as: `resolve()?.typeName() ?: fallbackClassName()`, might be also able to lift the return to the if/else